File: lib/python/opentoken/src/test/opentoken/io/csv/person_attributes_csv_reader_test.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import os
import tempfile

import pytest

from opentoken.attributes.general.record_id_attribute import RecordIdAttribute
from opentoken.attributes.person.first_name_attribute import FirstNameAttribute
from opentoken.attributes.person.last_name_attribute import LastNameAttribute
from opentoken.attributes.person.social_security_number_attribute import SocialSecurityNumberAttribute
from opentoken.io.csv.person_attributes_csv_reader import PersonAttributesCSVReader


class TestPersonAttributesCSVReader:
    """Test cases for PersonAttributesCSVReader."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.temp_file = tempfile.NamedTemporaryFile(mode='w', suffix='.csv', delete=False)
        self.temp_file_path = self.temp_file.name
        self.temp_file.close()

    def teardown_method(self):
        """Clean up after each test method."""
        if os.path.exists(self.temp_file_path):
            os.unlink(self.temp_file_path)

    def test_read_valid_csv(self):
        """Test reading a valid CSV file."""
        with open(self.temp_file_path, 'w', encoding='utf-8') as f:
            f.write("RecordId,SocialSecurityNumber,FirstName,LastName\n")
            f.write("1,123-45-6789,John,Doe\n")
            f.write("2,987-65-4321,Jane,Smith\n")

        with PersonAttributesCSVReader(self.temp_file_path) as reader:
            # Test first record
            first_record = next(reader)
            assert first_record[RecordIdAttribute] == "1"
            assert first_record[SocialSecurityNumberAttribute] == "123-45-6789"
            assert first_record[FirstNameAttribute] == "John"
            assert first_record[LastNameAttribute] == "Doe"

            # Test second record
            second_record = next(reader)
            assert second_record[RecordIdAttribute] == "2"
            assert second_record[SocialSecurityNumberAttribute] == "987-65-4321"
            assert second_record[FirstNameAttribute] == "Jane"
            assert second_record[LastNameAttribute] == "Smith"

            # Test no more records
            with pytest.raises(StopIteration):
                next(reader)

    def test_read_empty_csv(self):
        """Test reading an empty CSV file."""
        with open(self.temp_file_path, 'w', encoding='utf-8') as f:
            f.write("RecordId,SocialSecurityNumber,Name\n")

        with PersonAttributesCSVReader(self.temp_file_path) as reader:
            with pytest.raises(StopIteration):
                next(reader)

    def test_iterator_protocol(self):
        """Test iterator protocol."""
        with open(self.temp_file_path, 'w', encoding='utf-8') as f:
            f.write("RecordId,SocialSecurityNumber,FirstName,LastName\n")
            f.write("1,123-45-6789,John,Doe\n")

        with PersonAttributesCSVReader(self.temp_file_path) as reader:
            # Test that we can iterate
            for record in reader:
                assert record[RecordIdAttribute] == "1"
                assert record[SocialSecurityNumberAttribute] == "123-45-6789"
                assert record[FirstNameAttribute] == "John"
                assert record[LastNameAttribute] == "Doe"
                break

    def test_next(self):
        """Test next method."""
        with open(self.temp_file_path, 'w', encoding='utf-8') as f:
            f.write("RecordId,SocialSecurityNumber,FirstName,LastName\n")
            f.write("1,123-45-6789,John,Doe\n")

        with PersonAttributesCSVReader(self.temp_file_path) as reader:
            record = next(reader)
            assert record is not None
            assert record[RecordIdAttribute] == "1"
            assert record[SocialSecurityNumberAttribute] == "123-45-6789"
            assert record[FirstNameAttribute] == "John"
            assert record[LastNameAttribute] == "Doe"

    def test_close(self):
        """Test close method."""
        with open(self.temp_file_path, 'w', encoding='utf-8') as f:
            f.write("RecordId,SocialSecurityNumber,Name\n")
            f.write("1,123-45-6789,John Doe\n")

        reader = PersonAttributesCSVReader(self.temp_file_path)
        reader.close()

        # After closing, next should raise StopIteration
        with pytest.raises(ValueError):
            next(reader)

    def test_constructor_throws_io_exception(self):
        """Test constructor throws IOError for non-existent file."""
        invalid_file_path = "non_existent_file.csv"
        with pytest.raises(IOError):
            PersonAttributesCSVReader(invalid_file_path)

File: lib/java/opentoken/src/test/java/com/truveta/opentoken/io/csv/PersonAttributesCSVReaderTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.csv;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Map;
import java.util.NoSuchElementException;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.truveta.opentoken.attributes.Attribute;
import com.truveta.opentoken.attributes.general.RecordIdAttribute;
import com.truveta.opentoken.attributes.person.FirstNameAttribute;
import com.truveta.opentoken.attributes.person.LastNameAttribute;
import com.truveta.opentoken.attributes.person.SocialSecurityNumberAttribute;

class PersonAttributesCSVReaderTest {

    private File tempFile;

    @BeforeEach
    void setUp() throws IOException {
        tempFile = File.createTempFile("test_data", ".csv");
    }

    @AfterEach
    void tearDown() {
        if (tempFile.exists()) {
            tempFile.delete();
        }
    }

    @Test
    void testReadValidCSV() throws Exception {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile))) {
            writer.write("RecordId,SocialSecurityNumber,FirstName,LastName\n");
            writer.write("1,123-45-6789,John,Doe\n");
            writer.write("2,987-65-4321,Jane,Smith\n");
        }

        try (PersonAttributesCSVReader reader = new PersonAttributesCSVReader(tempFile.getAbsolutePath())) {
            assertTrue(reader.hasNext());

            Map<Class<? extends Attribute>, String> firstRecord = reader.next();
            assertEquals("1", firstRecord.get(RecordIdAttribute.class));
            assertEquals("123-45-6789", firstRecord.get(SocialSecurityNumberAttribute.class));
            assertEquals("John", firstRecord.get(FirstNameAttribute.class));
            assertEquals("Doe", firstRecord.get(LastNameAttribute.class));

            assertTrue(reader.hasNext());

            Map<Class<? extends Attribute>, String> secondRecord = reader.next();
            assertEquals("2", secondRecord.get(RecordIdAttribute.class));
            assertEquals("987-65-4321", secondRecord.get(SocialSecurityNumberAttribute.class));
            assertEquals("Jane", secondRecord.get(FirstNameAttribute.class));
            assertEquals("Smith", secondRecord.get(LastNameAttribute.class));

            assertFalse(reader.hasNext());
        }
    }

    @Test
    void testReadEmptyCSV() throws Exception {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile))) {
            writer.write("RecordId,SocialSecurityNumber,Name\n");
        }

        try (PersonAttributesCSVReader reader = new PersonAttributesCSVReader(tempFile.getAbsolutePath())) {
            assertFalse(reader.hasNext());
        }
    }

    @Test
    void testHasNext() throws Exception {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile))) {
            writer.write("RecordId,SocialSecurityNumber,FirstName,LastName\n");
            writer.write("1,123-45-6789,John,Doe\n");
        }

        try (PersonAttributesCSVReader reader = new PersonAttributesCSVReader(tempFile.getAbsolutePath())) {
            assertTrue(reader.hasNext());
            reader.next();
            assertFalse(reader.hasNext());
        }
    }

    @Test
    void testNext() throws Exception {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile))) {
            writer.write("RecordId,SocialSecurityNumber,FirstName,LastName\n");
            writer.write("1,123-45-6789,John,Doe\n");
        }

        try (PersonAttributesCSVReader reader = new PersonAttributesCSVReader(tempFile.getAbsolutePath())) {
            Map<Class<? extends Attribute>, String> record = reader.next();
            assertNotNull(record);
            assertEquals("1", record.get(RecordIdAttribute.class));
            assertEquals("123-45-6789", record.get(SocialSecurityNumberAttribute.class));
            assertEquals("John", record.get(FirstNameAttribute.class));
            assertEquals("Doe", record.get(LastNameAttribute.class));
        }
    }

    @Test
    void testClose() throws Exception {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile))) {
            writer.write("RecordId,SocialSecurityNumber,Name\n");
            writer.write("1,123-45-6789,John Doe\n");
        }

        PersonAttributesCSVReader reader = new PersonAttributesCSVReader(tempFile.getAbsolutePath());
        reader.close();
        assertThrows(NoSuchElementException.class, reader::next);
    }

    @Test
    void testConstructorThrowsIOException() {
        String invalidFilePath = "non_existent_file.csv";
        assertThrows(IOException.class, () -> new PersonAttributesCSVReader(invalidFilePath));
    }
}